bindings/c/test/unit/unit_tests.cpp

```diff
@@ -0,0 +1,1575 @@
+      fdb_check(wait_future(f6));
+      continue;
+    }
+
+	// Error can occur on any future.
+    fdb_error_t err1 = wait_future(f1);
+    if (err1) {
+      CHECK(err1 == 1032); // too_many_watches
+      break;
+    }
+
+    fdb_error_t err2 = wait_future(f2);
+    if (err2) {
+      CHECK(err2 == 1032); // too_many_watches
+      break;
+    }
+
+    fdb_error_t err3 = wait_future(f3);
+    if (err3) {
+      CHECK(err3 == 1032); // too_many_watches
+      break;
+    }
+
+    fdb_error_t err4 = wait_future(f4);
+    if (err4) {
+      CHECK(err4 == 1032); // too_many_watches
+      break;
+    }
+
+    // One of the watches should have triggered a too_many_watches failure.
+    CHECK(false);
+    break;
+  }
+
+  // Reset available number of watches.
+  max_watches = 10000;
+  fdb_check(fdb_database_set_option(db, FDB_DB_OPTION_MAX_WATCHES,
+                                    (const uint8_t *)&max_watches, 8));
+}
+
+TEST_CASE("fdb_transaction_watch") {
+  fdb::Transaction tr(db);
+
+  struct Context {
+    FdbEvent event;
+  };
+  Context context;
+
+  while (1) {
+    fdb::EmptyFuture f1 = tr.watch(KEY("foo"), KEYSIZE("foo"));
+    fdb::EmptyFuture f2 = tr.commit();
+
+    fdb_error_t err = wait_future(f2);
+    if (err) {
+      fdb::EmptyFuture f3 = tr.on_error(err);
+      fdb_check(wait_future(f3));
+      continue;
+    }
+
+    fdb_check(f1.set_callback(+[](FDBFuture *f1, void *param) {
+                                auto *context =
+                                    static_cast<Context *>(param);
+                                context->event.set();
+                              },
+                              &context));
+    break;
+  }
+
+  // Insert data for key "foo" to trigger the watch.
+  insert_data(db, create_data({ { "foo", "bar" } }));
+  context.event.wait();
+}
+
+TEST_CASE("fdb_transaction_cancel") {
+  // Cannot use transaction after cancelling it...
+  fdb::Transaction tr(db);
+  tr.cancel();
+  fdb::ValueFuture f1 = tr.get((const uint8_t *)"foo", 3, /* snapshot */ false);
+  CHECK(wait_future(f1) == 1025); // transaction_cancelled
+
+  // ... until the transaction has been reset.
+  tr.reset();
+  fdb::ValueFuture f2 = tr.get((const uint8_t *)"foo", 3, /* snapshot */ false);
+  fdb_check(wait_future(f2));
+}
+
+TEST_CASE("block_from_callback") {
+  fdb::Transaction tr(db);
+  fdb::ValueFuture f1 = tr.get((const uint8_t *)"foo", 3, /*snapshot*/ true);
+  struct Context {
+    FdbEvent event;
+    fdb::Transaction *tr;
+  };
+  Context context;
+  context.tr = &tr;
+  fdb_check(f1.set_callback(
+      +[](FDBFuture *f1, void *param) {
+            auto *context = static_cast<Context *>(param);
+            fdb::ValueFuture f2 = context->tr->get((const uint8_t *)"bar", 3,
+                                                /*snapshot*/ true);
+            fdb_error_t error = f2.block_until_ready();
+            if (error) {
+              CHECK(error == /*blocked_from_network_thread*/ 2025);
```

Comment:
I don't think this feature is implemented. Maybe disable for now and we can add it back when it gets merged into master?